datafusion/core/src/physical_plan/joins/hash_join.rs

```diff
@@ -781,22 +780,33 @@ pub fn build_equal_condition_join_indices(
     //     (5,1)
     //
     // With this approach, the lexicographic order on both the probe side and the build side is preserved.
+    let hash_map = build_hashmap.get_map();
+    let next_chain = build_hashmap.get_list();
     for (row, hash_value) in hash_values.iter().enumerate().rev() {
         // Get the hash and find it in the build index
 
         // For every item on the build and probe we check if it matches
         // This possibly contains rows with hash collisions,
         // So we have to check here whether rows are equal or not
-        if let Some((_, index)) = build_hashmap
-            .map
-            .get(*hash_value, |(hash, _)| *hash_value == *hash)
+        if let Some((_, index)) =
+            hash_map.get(*hash_value, |(hash, _)| *hash_value == *hash)
         {
             let mut i = *index - 1;
             loop {
-                build_indices.append(i);
+                let build_row_value = if let Some(offset) = deleted_offset {
```

Comment:
Wonder if this doesn't cause any regressions?

Reply:
I will investigate any potential performance issues

Reply:
The compiler should be able to hoist the check above the for loop, so the check should essentially be done only once in practice and not cause any perf issues. @metesynnada will verify this, but if it turns out the compiler can not do this automatically, we can do the hoisting manually (and unfortunately duplicate ~10 line the for loop code).

Reply:
```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ apache_main ┃ fork_sym-hashtable ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │    552.79ms │           550.33ms │     no change │
│ QQuery 2     │     72.21ms │            71.10ms │     no change │
│ QQuery 3     │    174.05ms │           171.05ms │     no change │
│ QQuery 4     │    115.42ms │           110.86ms │     no change │
│ QQuery 5     │    216.56ms │           214.26ms │     no change │
│ QQuery 6     │    107.64ms │           106.45ms │     no change │
│ QQuery 7     │    308.30ms │           307.71ms │     no change │
│ QQuery 8     │    247.63ms │           244.98ms │     no change │
│ QQuery 9     │    374.71ms │           368.62ms │     no change │
│ QQuery 10    │    295.30ms │           292.97ms │     no change │
│ QQuery 11    │     66.83ms │            61.84ms │ +1.08x faster │
│ QQuery 12    │    213.71ms │           207.88ms │     no change │
│ QQuery 13    │    315.16ms │           312.82ms │     no change │
│ QQuery 14    │    156.43ms │           150.26ms │     no change │
│ QQuery 15    │    103.82ms │           102.99ms │     no change │
│ QQuery 16    │     62.30ms │            59.98ms │     no change │
│ QQuery 17    │    277.65ms │           258.28ms │ +1.08x faster │
│ QQuery 18    │    569.40ms │           535.63ms │ +1.06x faster │
│ QQuery 19    │    321.54ms │           322.04ms │     no change │
│ QQuery 20    │    188.54ms │           179.86ms │     no change │
│ QQuery 21    │    450.69ms │           457.03ms │     no change │
│ QQuery 22    │     74.30ms │            72.68ms │     no change │
└──────────────┴─────────────┴────────────────────┴───────────────┘
```
I think there is only measurement noise here.

Reply:
Looks good, thanks :)